### PR TITLE
use a hash table in the expression formatter

### PR DESCRIPTION
**Pull request by kroening**
This replaces the linear-time if-then-else construct by a logarithmic hash
table.  The hash table could be extended at runtime.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/format_expr.cpp [hunk 1]
```diff
@@ -217,81 +217,145 @@ std::ostream &fallback_format_rec(std::ostream &os, const exprt &expr)
...
+  }
+
+  using formattert =
+    std::function<std::ostream &(std::ostream &, const exprt &)>;
+  using expr_mapt = std::unordered_map<irep_idt, formattert>;
+
+  expr_mapt expr_map;
+
+  /// find the formatter for a given expression
+  const formattert &find_formatter(const exprt &);
```
Comment (tautschnig) [inline src/util/format_expr.cpp:235]: Shouldn't this be `const`?

File: src/util/format_expr.cpp [hunk 2]
```diff
@@ -217,81 +217,145 @@ std::ostream &fallback_format_rec(std::ostream &os, const exprt &expr)
...
+
+private:
+  /// setup the expressions we can format
+  void setup();
```
Comment (tautschnig) [inline src/util/format_expr.cpp:239]: Possibly a matter of taste: Is there value in having this extra method as opposed to just doing all work directly in the constructor?

**Review by tautschnig — APPROVED** [review]

File: src/util/format_expr.cpp (hunk 2 above)
Comment (kroening) [reply src/util/format_expr.cpp:239]: Not a whole lot, it may convey a bit of information to the reader.

File: src/util/format_expr.cpp (hunk 1 above)
Comment (kroening) [reply src/util/format_expr.cpp:235]: As-is now yes, but was contemplating to generate formatters on the fly.

**Pull request merged**